poppy/accel_math.py

```diff
@@ -64,16 +88,220 @@ def _exp(x):
+        do_fft = pyfftw.interfaces.numpy_fft.fft2 if forward else pyfftw.interfaces.numpy_fft.ifft2
+        if normalization is None:
+            normalization = 1./wavefront.shape[0] if forward else wavefront.shape[0]
+
+
+        if (wavefront.shape, FFT_direction) not in _FFTW_INIT:
+            # The first time you run FFTW to transform a given size, it does a speed test to
+            # determine optimal algorithm that is destructive to your chosen array.
+            # So only do that test on a copy, not the real array:
+            _log.info("Measuring pyfftw optimal plan for %s, direction=%s" % (
+                str(wavefront.shape), FFT_direction))
+
+            pyfftw.interfaces.cache.enable()
+            pyfftw.interfaces.cache.set_keepalive_time(30)
+
+            test_array = np.zeros(wavefront.shape)
+            test_array = do_fft(test_array, overwrite_input=True, planner_effort='FFTW_MEASURE',
+                                threads=multiprocessing.cpu_count())
+
+            _FFTW_INIT[(wavefront.shape, FFT_direction)] = True
+
+        wavefront = do_fft(wavefront, overwrite_input=True, planner_effort='FFTW_MEASURE',
+                                threads=multiprocessing.cpu_count())
+    else: # Basic numpy FFT
+        do_fft =  np.fft.fft2 if forward else np.fft.ifft2
+        if normalization is None:
+            normalization = 1./wavefront.shape[0] if forward else wavefront.shape[0]
+        wavefront = do_fft(wavefront)
+
+    if forward and fftshift:
+        wavefront = _fftshift(wavefront)
+
+    wavefront *= normalization
+
+    return wavefront
+
+
+
+def ispowerof2(num):
+    """ Is this number a power of 2?"""
+    # see http://code.activestate.com/recipes/577514-chek-if-a-number-is-a-power-of-two/
+    return (num & (num-1) == 0)
+
+if _OPENCL_AVAILABLE:
+    def get_opencl_context():
+        """ Create, save, and retrieve OpenCL handles to the GPU """
+        if len(_OPENCL_STATE) == 0:
+            platforms = pyopencl.get_platforms()
+            if len(platforms) == 1:
+                _OPENCL_STATE['platform'] = platforms[0]
+            else:
+                raise RuntimeError("OpenCL code needs update for multiple platforms")
+            gpus = _OPENCL_STATE['platform'].get_devices(device_type=pyopencl.device_type.GPU)
+            if len(gpus) == 1:
+                device = gpus[0]
+                _OPENCL_STATE['device'] = device
+            else:
+                #raise RuntimeError("OpenCL code could not uniquely identify which device to use as GPU")
+                device = gpus[1]
+                _log.warning('Caution - hard coded use of gpu #1 if > 1 GPUs present')
+            context = pyopencl.Context(devices=[device])
+            queue = pyopencl.CommandQueue(context)
+
+            _OPENCL_STATE['context'] = context
+            _OPENCL_STATE['queue'] = queue
+        return (_OPENCL_STATE['context'], _OPENCL_STATE['queue'])
+
+
+
+
+
 if  _USE_CUDA:
```

Comment:
TO

Reply:
Not sure what this comment means?

Reply:
it was supposed to be TO DO: figure out how to rewrite this in OpenCL

Reply:
Right!  And the lack of that is a big part of why I wrote `fft_2d` to prioritize CUDA rather than OpenCL if both are available...